From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Igor Sheludko <isheludko@gmail.com>
Date: Tue, 30 Jun 2026 15:38:51 +0200
Subject: [v8] Propagate v8_cppgc_microtask_queue flag to common.gypi

v8_cppgc_microtask_queue affects the public ABI of v8::MicrotaskQueue.
When enabled, it alters the class hierarchy (inheriting from cppgc::GarbageCollected)
and changes the return type of MicrotaskQueue::New().

This flag must be propagated to GYP to ensure that native addons built via
node-gyp are compiled with the identical V8 memory layout and method signatures
as the Node.js executable, preventing silent ABI breakages and segfaults.

Backports https://github.com/v8/node/commit/cb6ab8108df3cea31c673cb8c721afeffad99540
Refs https://chromium-review.googlesource.com/c/v8/v8/+/7991718

Will be removed when Node.js bumps to version of V8 that contains
the API change.

diff --git a/common.gypi b/common.gypi
index 4c5e8ee2a6e97d0877550a836cf371f800e33bc5..483a145d65aa9f3ba836e7abaf403aa45cbe8fb1 100644
--- a/common.gypi
+++ b/common.gypi
@@ -92,6 +92,7 @@
     'v8_enable_zone_compression%': 0,
     'v8_use_perfetto': 0,
     'tsan%': 0,
+    'v8_cppgc_microtask_queue%': 0,
 
     'using_electron_config_gypi%': 0,
 
@@ -560,6 +561,9 @@
       ['v8_enable_direct_local == 1', {
         'defines': ['V8_ENABLE_DIRECT_LOCAL',],
       }],
+      ['v8_cppgc_microtask_queue == 1', {
+        'defines': ['V8_CPPGC_MICROTASK_QUEUE',],
+      }],
       ['OS == "win"', {
         'defines': [
           'WIN32',
diff --git a/configure.py b/configure.py
index f7577d93244e4543db732f7e4c1f0db14f0565b2..86aea91a11eb9bf39247ccf90479542aa7558319 100755
--- a/configure.py
+++ b/configure.py
@@ -762,6 +762,14 @@ parser.add_argument('--experimental-pointer-compression-shared-cage',
     default=None,
     help='[Experimental] Use V8 pointer compression with shared cage (requires --experimental-enable-pointer-compression)')
 
+parser.add_argument('--v8-enable-cppgc-microtask-queue',
+    action='store_true',
+    dest='v8_enable_cppgc_microtask_queue',
+    default=None,
+    help='Allocate V8 microtask queues with cppgc. This must match how the '
+         'embedded V8 was built, since it affects the ABI exposed through the '
+         'public V8 headers.')
+
 parser.add_argument('--v8-options',
     action='store',
     dest='v8_options',
@@ -1995,6 +2003,7 @@ def configure_v8(o, configs):
   o['variables']['v8_enable_pointer_compression_shared_cage'] = 1 if options.pointer_compression_shared_cage else 0
   o['variables']['v8_enable_external_code_space'] = 1 if options.enable_pointer_compression else 0
   o['variables']['v8_enable_31bit_smis_on_64bit_arch'] = 1 if options.enable_pointer_compression else 0
+  o['variables']['v8_cppgc_microtask_queue'] = 1 if options.v8_enable_cppgc_microtask_queue else 0
   o['variables']['v8_enable_extensible_ro_snapshot'] = 0
   o['variables']['v8_trace_maps'] = 1 if options.trace_maps else 0
   o['variables']['node_use_v8_platform'] = b(not options.without_v8_platform)
